worker: align MessagePort message handling more with Web

addaleax opened the pull request:
This aligns `MessagePort`s more with the web API and resolved https://github.com/nodejs/node/issues/26463.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented:
/cc @nodejs/workers

CI: https://ci.nodejs.org/job/node-test-pull-request/21299/

gireeshpunathil commented on lib/internal/worker/io.js:
```diff
@@ -78,7 +78,6 @@ Object.defineProperty(MessagePort.prototype, 'onmessage', {
       MessagePortPrototype.start.call(this);
     } else {
       this.unref();
-      stopMessagePort(this);
```
this behavioral difference : between `onmessage` and `on('message')` - should we document this? it may be apparent to some class of users, but not so for others?

addaleax replied:
We currently don’t document `onmessage` and its behaviour ourselves (but MDN does).

I’ve added a paragraph about this to the documentation, though.

addaleax commented on test/parallel/test-worker-message-port-onmessage-stop.js:
```diff
@@ -0,0 +1,30 @@
+
+{
+  const { port1, port2 } = new MessageChannel();
+  port1.onmessage = common.mustCall(({ data }) => {
```
You mean the ones around `data`? Yes, that’s destructuring the argument here

addaleax commented on test/parallel/test-worker-message-port-onmessage-stop.js:
```diff
@@ -0,0 +1,30 @@
+    port2.postMessage('world');
+
+    // Give some event loop ticks for the 'world' message to be discarded.
+    tick(10, () => {
```
I’m not sure I understand what you are saying – what would wrongly activating (= calling?) `.onmessage` mean?

gireeshpunathil replied:
`port1.onmessage = null;`  // port1 is ditached from messages

`port2.postMessage('world');`

What we want to check is `port1.onmessage` should not be called henceforth. But if it is being called by error, we don't have a means to catch it?

addaleax replied:
@gireeshpunathil How do you test that something that is not a function isn’t called?

We can only check that there’s nothing in the queue after a few ticks, and that’s what setting `onmessage = common.mustNotCall()` is there for

gireeshpunathil replied:
@addaleax - I see your point. What I thought was something like this:
```js
  const { port1, port2 } = new MessageChannel();
  port1.onmessage = common.mustCall(({ data }) => {
    assert.strictEqual(data, 'hello');
    port1.onmessage = null;
    port2.postMessage('world');
  }, 1);
  port2.postMessage('hello');
```

Given that we are dealing with only one thread here, and the port1 being detached from its message channel would happen synchronously, we should see the result immediately, and the subsequent emit from port2 itself could be subjected for the assert test? assert that we don't call the handler more than once?

addaleax replied:
> assert that we don't call the handler more than once?

Still not quite sure – we do use `common.mustCall()` currently for the “outermost” `onmessage`? Can you provide an example of changes to the test that you would like to see?

gireeshpunathil replied:
This snippet itself is what I am proposing. if the intent of the test is to: `Make sure that messages start being received when .onmessage is first set, but when it is unset, messages end up being discarded.` , then won't this sequence test it reliably?

Once the `onmessage` is unset, the only possible error that can occur is it is bound to the function somehow, and the callback (outer onmessage, in fact now the only one in the program) will be invoked, and that will be captured by the common.mustCall that mandates only one invocation?

```js
const { port1, port2 } = new MessageChannel();
port1.onmessage = common.mustCall(({ data }) => {
  assert.strictEqual(data, 'hello');
  port1.onmessage = null;
  port2.postMessage('world');
}, 1);
port2.postMessage('hello');
```

addaleax replied:
@gireeshpunathil There is already a `common.mustCall()` in the test in this PR for the outer onmessage, and another error condition which we want to check for would be that the second message is just queued up and not discarded (as it would be before this patch), which wouldn’t be caught if we leave `onmessage` as `null` until the test ends?

gireeshpunathil replied:
ah, ok! I realize it now, thanks.

Trott commented on doc/api/worker_threads.md:
```diff
@@ -203,6 +203,11 @@ input of [`port.postMessage()`][].
 to `postMessage()` and no further arguments.
 
+When there are `'message'` event listeners and `.start()` is not called
+explicitly, and the Web-based `.onmessage` property is not used, no messages
```
`web-based`?

Trott replied:
(Maybe `web-baed` should be removed? It seems likely to be misunderstood. I *think* we just mean "modeled on the analogous web-based API" but people are likely to interpret as more like...something to do with the web directly? That's what I did when I read this for a few seconds.

Trott replied:
(Both comments above are non-blocking.)

addaleax replied:
I think removing `web-based` would be a bit confusing, given that we don’t currently document `onmessage` ourselves.

chjj commented:
I'd like to look at this more closely.

To keep things clear, the current behavior on master is:

- `.on('message', fn)` with no listeners -> reference and start
- `.off('message', fn)` with one listener -> dereference and stop
- `.onmessage = fn` -> reference and start
- `.onmessage = null` -> dereference and stop

This PR changes the behavior to:

- `.on('message', fn)` with no listeners -> reference and start (unchanged)
- `.off('message', fn)` with one listener -> dereference and stop (unchanged)
- `.onmessage = fn` -> reference and start (unchanged)
- `.onmessage = null` -> dereference, _do not_ stop (__changed__)

I believe the browser behavior should be:

- `.on('message', fn)` with no listeners -> reference, _do not_ start
- `.off('message', fn)` with one listener -> dereference, _do not_ stop
- `.onmessage = fn` -> reference and start
- `.onmessage = null` -> dereference, _do not_ stop

I'll have to write up some code to confirm that the latter is actually the browser behavior. It might take me a second.

If that is the behavior, I there's one more thing to consider: the browser doesn't have references since it's an ever-living event loop (I'm pretty sure workers can't exit gracefully in-browser). That said, we could maybe consider removing the magical referencing behavior of `.on('message')` and `.off('message')` altogether since it doesn't actually have anything to do with starting or stopping the port anymore. This would mean that the `parentPort` would have to be started by default (I believe it is in the browser, since `self.start()` [does not exist], but I'll have to test this as well.

It's probably worth discussing a bit more before anything is merged.

[does not exist]: https://developer.mozilla.org/en-US/docs/Web/API/DedicatedWorkerGlobalScope/onmessage

addaleax commented:
@chjj I think you’re correct about the behaviour everywhere here.

The behaviour for the Node.js-style APIs was picked because it made the most sense to me, and since the browser has neither `.on` nor `.off` methods (and the browser lacks `EventEmitter` as an API and Node.js lacks `EventTarget` as an API in general), it didn’t seem like a full match was feasible or desirable.

For example, the “event” that the `onmessage` function receives in Node.js is not really close to being a full `MessageEvent`.

chjj commented:
@addaleax, yeah, the more I think about it, the way node does it makes more sense than the browser. In my own worker compatibility module, I actually hack the browser into behaving like node since it seemed to make more sense in the end. 

Personally, I'd be fine with having things stay the way they are and simply document the difference (i.e. node.js will automatically start and stop the buffering of the port).

Anyway, I'm going to tinker with the browser more to confirm the exact behavior. Right now, I suspect that `addEventListener('message')` has no magic behavior and that `DedicatedWorkerGlobalScope` is akin to a `MessagePort` started by default.

chjj commented:
I've confirmed that MDN was correct about `port.addEventListener('message')`: it does not start the port in the browser.

I've also confirmed that [DedicatedWorkerGlobalScope] behaves like a `MessagePort` which is started by default (i.e. you will miss `self.onmessage` events if you don't bind the listener right away). I don't think it's worth starting the parentPort by default just for browser compat, since there's probably some legitimate use cases for not doing that in node.js.

I think the third behavior I described in my first post is probably the way to go if we want browser compat. That said, it might be worth just keeping things the way they are. I do think the current node.js behavior is much saner than the browser. I think my original issue may have been bikeshedding a bit. Would love to hear your thoughts, @addaleax.

[DedicatedWorkerGlobalScope]: https://developer.mozilla.org/en-US/docs/Web/API/DedicatedWorkerGlobalScope

addaleax commented:
@chjj So… where would you stand on this PR itself? I agree with you in that the browser behaviour is odd, but I also see value in explicitly aligning with the Web spec here.

Then again, I’m not even sure that the difference is guaranteed to be observable, because in theory receiving a message can take an arbitrary amount of time (and thus just “randomly” happen when a new `onmessage` listener has been attached).

I’m personally happy to close this and just document the difference, as you suggested above?

chjj commented:
@addaleax, as far as the PR, I'd say the `Cannot send data on closed port` error should still be removed, but no changes should be made to the `onmessage` or `.on('message')` start/stop behavior. I think the current node behavior is good. Documenting the differences from the browser is probably sufficient.

addaleax commented:
@chjj Okay, I think we’re in agreement – I’ve removed the changes for `.onmessage` and added docs for the divergence from the browser.

@gireeshpunathil @benjamingr Do you want to take another look?

gireeshpunathil commented:
still LGTM.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/21404/

addaleax commented:
Landed in a52aedeae08b57daedae56fd818be09df5f32e86, 137d3d2734a6f6e891ebab68ec6da16c8a6b397c
